sys/psa_crypto/psa_crypto_location_dispatch.c

```diff
@@ -0,0 +1,440 @@
+    const psa_drv_se_t *drv;
+    psa_drv_se_context_t *drv_context;
+    psa_key_slot_number_t *slot_number = psa_key_slot_get_slot_number(slot);
+
+    uint8_t *pubkey_data = NULL;
+    size_t *pubkey_data_len = NULL;
+
+    psa_get_public_key_data_from_key_slot(slot, &pubkey_data, &pubkey_data_len);
+
+    if (psa_get_se_driver(attributes->lifetime, &drv, &drv_context)) {
+        if (drv->key_management == NULL || drv->key_management->p_generate == NULL) {
+            return PSA_ERROR_NOT_SUPPORTED;
+        }
+
+        status = drv->key_management->p_generate(drv_context, *slot_number, attributes, pubkey_data,
+                                               *pubkey_data_len, pubkey_data_len);
+        if (status != PSA_SUCCESS) {
+            /* In case anything goes wrong, free the key slot for reuse. */
+            psa_se_drv_data_t *driver = psa_get_se_driver_data(attributes->lifetime);
+            psa_status_t abort_status = drv->key_management->p_destroy(drv_context, driver->ctx.internal.persistent_data, *slot_number);
+            return abort_status == PSA_SUCCESS ? status : abort_status;
+        }
+        return PSA_SUCCESS;
+    }
+#endif /* MODULE_PSA_SECURE_ELEMENT */
+
+    return psa_algorithm_dispatch_generate_key(attributes, slot);
+}
+
+psa_status_t psa_location_dispatch_import_key( const psa_key_attributes_t *attributes,
+                                               const uint8_t *data, size_t data_length,
+                                               psa_key_slot_t *slot, size_t *bits)
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_key_location_t location = PSA_KEY_LIFETIME_GET_LOCATION(attributes->lifetime);
+    uint8_t *key_data = NULL;
+    size_t *key_bytes = NULL;
+
+    psa_get_key_data_from_key_slot(slot, &key_data, &key_bytes);
+
+#if IS_USED(MODULE_PSA_SECURE_ELEMENT)
+    const psa_drv_se_t *drv;
+    psa_drv_se_context_t *drv_context;
+    psa_key_slot_number_t *slot_number = psa_key_slot_get_slot_number(slot);
+
+    if (psa_get_se_driver(attributes->lifetime, &drv, &drv_context)) {
+        if (drv->key_management == NULL || drv->key_management->p_import == NULL) {
+            return PSA_ERROR_NOT_SUPPORTED;
+        }
+        *bits = 0;
+
+        status = drv->key_management->p_import(drv_context, *slot_number,
+                                               attributes, data,
+                                               data_length, bits);
+        if (status != PSA_SUCCESS) {
+            /* In case anything goes wrong, free the key slot for reuse. */
+            psa_se_drv_data_t *driver = psa_get_se_driver_data(attributes->lifetime);
+            psa_status_t abort_status = drv->key_management->p_destroy(drv_context, driver->ctx.internal.persistent_data, *slot_number);
+            return abort_status == PSA_SUCCESS ? status : abort_status;
+        }
+        return PSA_SUCCESS;
+    }
+#endif /* MODULE_PSA_SECURE_ELEMENT */
+
+    switch (location) {
```

Comment:
Why a switch?

Reply:
looks like future proofing; But shouldn't a secure Element be a key-storage-`location` ?

Reply:
Yes, this is future proofing. We know for sure that there are boards that have hardware protected key slots (I have a board here that provides this, it's just not supported, yet). For those we'll define additional locations.

Secure elements are also locations, but they are handled separately (above the switch).